cli/src/cmd/forge/verify.rs

```diff
@@ -152,42 +209,56 @@ impl VerifyArgs {
         } else {
-            standard_json_source(self, &project, &contract_path)?
+            self.standard_json_source(&project, &contract_path, &compiler_version)?
         };
 
+        let compiler_version = ensure_solc_build_metadata(compiler_version).await?;
+        let compiler_version = format!("v{}", compiler_version);
         let mut verify_args =
-            VerifyContract::new(self.address, contract_name, source, self.compiler_version.clone())
+            VerifyContract::new(self.address, contract_name, source, compiler_version)
                 .constructor_arguments(self.constructor_args.clone())
                 .code_format(code_format);
 
         if code_format == CodeFormat::SingleFile {
             verify_args = if let Some(optimizations) = self.num_of_optimizations {
-                verify_args.optimization(true).runs(optimizations)
+                verify_args.optimized().runs(optimizations as u32)
+            } else if config.optimizer {
+                verify_args.optimized().runs(config.optimizer_runs.try_into()?)
             } else {
-                verify_args.optimization(false)
-            };
+                verify_args.not_optimized()
+            }
         }
 
         Ok(verify_args)
     }
 
-    /// Parses the [Version] from the provided compiler version
-    ///
-    /// All etherscan supported compiler versions are listed here <https://etherscan.io/solcversions>
-    ///
-    /// **Note:** this is only for local compilation as a dry run, therefore this will return a
-    /// sanitized variant of the specific version so that it can be installed. This is merely
-    /// intended to ensure the flattened code can be compiled without errors.
-    ///
-    /// # Example
-    ///
-    /// the `compiler_version` `v0.8.7+commit.e28d00a7` will be returned as `0.8.7`
-    fn sanitized_solc_version(&self) -> eyre::Result<Version> {
-        let v: Version = self.compiler_version.trim_start_matches('v').parse()?;
-        Ok(Version::new(v.major, v.minor, v.patch))
+    /// Parse the compiler version.
+    /// The priority desc:
+    ///     1. Through CLI arg `--compiler-version`
+    ///     2. `solc` defined in foundry.toml  
+    fn compiler_version(&self, config: &Config) -> eyre::Result<Version> {
+        if let Some(ref version) = self.compiler_version {
+            return Ok(version.trim_start_matches('v').parse()?);
+        }
+
+        if let Some(ref solc) = config.solc {
+            match solc {
+                SolcReq::Version(version) => return Ok(version.to_owned()),
+                SolcReq::Local(solc) => {
+                    if solc.is_file() {
+                        return Ok(Solc::new(solc).version()?);
+                    }
+                }
+            }
+        }
+
+        eyre::bail!("Compiler version has to be set in `foundry.toml`. If the project were not deployed with foundry, specify the version through `--compiler-version` flag.")
```

Comment:
Should there be a 3rd priority, instead of erroring, which is "use whatever solc version + optimizer settings were used during last compilation"?

e.g from DS Test under `cache/solidity-files-cache.json`
```json
"artifacts": {
        "DSTest": {
          "0.8.13+commit.abaa5c0e.Darwin.appleclang": "test.sol/DSTest.json"
        }
```

Reply:
good point, no need to get the auto-detect version then